### ISSUE TITLE
Blame command has stopped working; IndexError whenever it tries to run

### ISSUE THREAD

**Issue by gparyani** [Charcoal-SE/SmokeDetector#5120]
The `!!/blame` command doesn't seem to work anymore. Using that command now results in an error message:

> I hit an error while trying to run that command; run `!!/errorlogs` for details.

When running the `!!/errorlogs` command, I get the following error log:

```python
### IndexError on teward/Osiris at 2020-11-15T08:42:20Z: list index out of range
File "/home/smokey/SmokeDetector/chatcommunicate.py", line 461, in __call__
result = self.__func__(*processed_args)
File "/home/smokey/SmokeDetector/chatcommands.py", line 615, in blame
unlucky_victim = msg._client.get_user(random.choice(msg.room.get_current_user_ids()))
File "/home/smokey/.local/lib/python3.6/site-packages/chatexchange/rooms.py", line 154, in get_current_user_ids
return self._client._br.get_current_user_ids_in_room(self.id)
File "/home/smokey/.local/lib/python3.6/site-packages/chatexchange/browser.py", line 681, in get_current_user_ids_in_room
return [user_id for (user_id, name) in self.get_current_users_in_room(room_id)]
File "/home/smokey/.local/lib/python3.6/site-packages/chatexchange/browser.py", line 671, in get_current_users_in_room
users_js = re.compile(r"(?s)CHAT\.RoomUsers\.initPresent\(\[.+\]\);").findall(script_text)[0]

### IndexError on teward/Osiris at 2020-11-15T08:42:03Z: list index out of range
File "/home/smokey/SmokeDetector/datahandling.py", line 520, in get_user_names_on_notification_list
current_users = client._br.get_current_users_in_room(room_id)
File "/home/smokey/.local/lib/python3.6/site-packages/chatexchange/browser.py", line 671, in get_current_users_in_room
users_js = re.compile(r"(?s)CHAT\.RoomUsers\.initPresent\(\[.+\]\);").findall(script_text)[0]
```

Seems like a list index is off somewhere.

Can this please be fixed?

**Comment by DavidPostill** [comment]
Note: Testing in the Tavern has shown that it is working for *some* users.

**Comment by rschrieken** [comment]
Is an upstream problem.

The `initPresent` used to be an Javascript call that took an array of userids in the page. The RegEx on line 671 finds and then returns the "array" as one string. In the latest version initPresent no longer is seeded with that array, so nothing is returned and that unexpected and not catered for.

What should be done is looking for a div with a class js-present and that div will have an attribute data-users containing the info you need. 

below is what that content looks like. Once htlmdecoded I expect the structure to be still the same so from [line 673](https://github.com/Charcoal-SE/ChatExchange/blob/master/chatexchange/browser.py#L673) onwards no code changes are needed.

Example from the Tavern room on MSE chat

``` 
<div class="js-present" data-users="[{&quot;id&quot;:267099,&quot;name&quot;:&quot;DavidPostill&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/gNepR.jpg?s=128&amp;g=1&quot;,&quot;reputation&quot;:7507,&quot;last_post&quot;:1605430681,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:377214,&quot;name&quot;:&quot;Sonic the K-Day Hedgehog&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/L9bAV.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:57755,&quot;last_post&quot;:1605430590,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:294691,&quot;name&quot;:&quot;Mithical&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/ramcj.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:38702,&quot;last_post&quot;:1605430562,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:true},{&quot;id&quot;:266345,&quot;name&quot;:&quot;SmokeDetector&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/7cmCt.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:101,&quot;last_post&quot;:1605430546,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:135565,&quot;name&quot;:&quot;Journeyman Geek&quot;,&quot;email_hash&quot;:&quot;e88cb2b3858ca0df735806f1a8476b23&quot;,&quot;reputation&quot;:96994,&quot;last_post&quot;:1605423552,&quot;is_moderator&quot;:true,&quot;is_owner&quot;:true},{&quot;id&quot;:152859,&quot;name&quot;:&quot;Shadow 10 Years Wizard&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/Zn7ED.gif?s=128&amp;g=1&quot;,&quot;reputation&quot;:144543,&quot;last_post&quot;:1605422668,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:158100,&quot;name&quot;:&quot;rene&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/qj33e.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:64153,&quot;last_post&quot;:1605370820,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:true},{&quot;id&quot;:194047,&quot;name&quot;:&quot;John Dvorak&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/PObJM.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:969,&quot;last_post&quot;:1605370317,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:280934,&quot;name&quot;:&quot;M.A.R.&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/dCyld.jpg?s=128&amp;g=1&quot;,&quot;reputation&quot;:7695,&quot;last_post&quot;:1605340465,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:186472,&quot;name&quot;:&quot;πάντα ῥεῖ&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/6wNQa.jpg?s=128&amp;g=1&quot;,&quot;reputation&quot;:13352,&quot;last_post&quot;:1605307889,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:165261,&quot;name&quot;:&quot;Ryan M&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/GKbQj.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:2387,&quot;last_post&quot;:1605261275,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:382836,&quot;name&quot;:&quot;Rubiksmoose&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/PiuWk.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:6462,&quot;last_post&quot;:1605235554,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:true},{&quot;id&quot;:262823,&quot;name&quot;:&quot;ArtOfCode&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/4aINw.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:34052,&quot;last_post&quot;:1605226394,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:true},{&quot;id&quot;:208518,&quot;name&quot;:&quot;JNat&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/5BO78.gif?s=128&amp;g=1&quot;,&quot;reputation&quot;:16317,&quot;last_post&quot;:1604936510,&quot;is_moderator&quot;:true,&quot;is_owner&quot;:true},{&quot;id&quot;:241919,&quot;name&quot;:&quot;Meta Andrew T.&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/Ibk3L.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:6295,&quot;last_post&quot;:1604565422,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:268944,&quot;name&quot;:&quot;Mark&quot;,&quot;email_hash&quot;:&quot;!https://i.stack.imgur.com/siang.png?s=128&amp;g=1&quot;,&quot;reputation&quot;:3561,&quot;last_post&quot;:1604534295,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:147015,&quot;name&quot;:&quot;ThiefMaster&quot;,&quot;email_hash&quot;:&quot;bc66db28aa7c3fa8e4ce4238b696dee9&quot;,&quot;reputation&quot;:15498,&quot;last_post&quot;:1604159177,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false},{&quot;id&quot;:150235,&quot;name&quot;:&quot;Tim Stone&quot;,&quot;email_hash&quot;:&quot;f51eafdfb4e45982b989365d3c461339&quot;,&quot;reputation&quot;:57009,&quot;last_post&quot;:1594143877,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false}]" data-joining="{&quot;id&quot;:0,&quot;name&quot;:&quot;&quot;,&quot;reputation&quot;:0,&quot;last_post&quot;:0,&quot;is_moderator&quot;:false,&quot;is_owner&quot;:false}"></div>
```

**Comment by tripleee** [comment]
FWIW this is probably also the reason we see repeated IndexErrors when Smokey tries to get the list of users in a room. The error is trapped but gets captured in `errors.log`.

**Comment by makyen** [comment]
Sorry to have resulted in other people doing additional work. I thought I'd pushed a fix for this to the Charcoal ChatExchange repository, but I'd actually only pushed it to my fork of that repository.

@rschrieken What I found was effectively identical to what you identified. The only real differences in implementation are that I chose to let the exceptions propagate up to the calling code, rather than return an empty list. I also left in the code for the prior scraping as a backup, in case SE reverts their change. In this case, reverting isn't all that likely, so leaving that in is of only marginal utility.

This was fixed in [7468d91](https://github.com/Charcoal-SE/ChatExchange/commit/7468d91e98bbecd95693677e5694555e35238284). [PR#160](https://github.com/Manishearth/ChatExchange/pull/160) has been submitted to the [main ChatExchange repository](https://github.com/Manishearth/ChatExchange). For it to be fixed for SmokeDetector instances, the instance runners will need to update dependency code. The other two instance runners [have been pinged](https://chat.stackexchange.com/transcript/message/56162493#56162493) to request that they do so.
